### DISCUSSION TITLE
Grouping classes by feature, not technical concern - what about templates?

### DISCUSSION THREAD

**Discussion by benr77** [symfony/symfony#47952] [Q&A]
So recently I've become a fan of grouping things together based on the feature they provide, rather than the type of class.

Symfony standard directory structure is something like the following:

```
.
├── Controller
│   ├── CreateProjectController.php
│   └── CreateUserController.php
├── Form
│   ├── CreateProjectType.php
│   └── CreateUserType.php
├── Serializer
│   ├── ProjectSerializer.php
│   └── UserNormalizer.php
└── Validator
    ├── ProjectValidator.php
    └── UserIsEligibleValidator.php
```

Now I like structuring things by feature like this:

```
.
└── Feature
    ├── CreateProject
    │   ├── CreateProjectController.php
    │   ├── CreateProjectType.php
    │   ├── ProjectSerializer.php
    │   └── ProjectValidator.php
    ├── CreateUser
    │   ├── CreateUserController.php
    │   ├── CreateUserType.php
    │   ├── UserIsEligibleValidator.php
    │   └── UserNormalizer.php
    └── SomeOtherFeature etc
```

Whilst this is a very simplistic example, you get the idea. In my real apps, where I have layered/hexagonal architecture, I'm currently only really doing this grouping in the UI layer at the moment, where grouping by feature feels very natural.

However, a pain point I've really noticed is when I refactor to the new structure in my `src` directory, I'm then having to follow-up and reorganise all the Twig templates in the `templates` directory. The templates feel a long way from where they are used in the controllers.

So seeing as templates are intrinsically linked with controller classes, why am I not grouping the templates along with the controller and form type classes? This can easily be achieved by configuring Twig to look in an additional location for templates:

```yaml
# config/packages/twig.yaml
twig:
    paths:
        'src/Feature/': 'Feature'
        'templates': ~
```

```
.
└── Feature
    └── CreateProject
        ├── CreateProjectController.php
        ├── CreateProjectType.php
        └── tpl
            └── create-project.html.twig   
```

Then you can use `$this->render('@Feature/CreateProject/tpl/create-project.html.twig')` in the controller.

Any common templates such as `layout.html.twig` can of course continue to live in the normal `templates` directory.

I'm not aware of any downsides such as performance problems, as all the Twig locations get worked out during cache warmup, so it's more a case of personal preference and ease of use.

I'm interested to see what the Symfony community thinks of this approach, as I'm pretty sure it's quite unconventional. Is it blasphemous to mix template files with PHP files? I guess it's a similar philosophy as placing your unit tests in the `src` directory (usually inside a `Tests` subdirectory of the production code) so that they are close to the code too.

So what does everyone think? Am I mad? Cheers!

**Comment by MatTheCat** [comment]
Many people would tell you this is blasphemous because they see defaults as the only way to go.

But frameworks should adapt to your needs, not the only way around. As Symfony embraces this philosophy there is absolutely nothing wrong moving from the defaults to something you find more convenient.

So you're not mad. Or is everyone? 🤔

**Comment by benr77** [comment]
Yes, not for a moment am I suggesting this is relevant for a default Symfony installation. But I have a large app with 120k lines of PHP, and details like this really start to matter. 

The `templates` directory is a huge dumping ground for the entire app and it's increasingly hard to keep organised. Splitting things up this way does seem prudent....

**Comment by ahundiak** [comment]
I suspect your approach is more common than you might think.  I have been using a similar feature/sub-feature layout since the days of 2.x.  If I have a class named UserController then I can be pretty sure that it is a controller without it being in a directory full of other controllers.

But here is something that is really blasphemous: I hardly ever use twig since twig templates are just flat procedural files with their own `autoloading` system.  I use Template classes which, while a bit more verbose than twig templates, are in fact actual classes with all the advantages those entail.  As a result, my Template classes happily coexist in the same directories as my controller or form or whatever classes.
